### ISSUE TITLE
How chose NetworkRegularizer to different CNN model?

### ISSUE THREAD

**Issue by Aspirinkb** [google-research/morph-net#55]
There are two types of `OpRegularizer`, `GammaL1Regularizer` and `GroupLassoRegularizer`. The classes `GammaFlopsRegularizer` and `GroupLassoFlopsRegularizer` in `flop_regularizer.py` implements two `NetworkRegularizer` targeting FLOPs. Now I have a ResNet model, how can I set regularizer? Does `GammaFlopsRegularizer` alone work well? Or shall I use `GammaFlopsRegularizer` and `GroupLassoFlopsRegularizer` together to train the model?

**Comment by Aspirinkb** [comment]
In the paper, 'MorphNet: Fast & Simple Resource-Constrained Structure Learning of Deep Networks', *4.4. Preserving the Network Topology* part said that:   
>DNNs in computer vision applications often have residual (skip) connections... To avoid these changes in the network topology, we group all neurons that are tied in skip connections via a Group LASSO   

So I think the `GammaFlopsRegularizer` and `GroupLassoFlopsRegularizer` should work together to  train a ResNet. Is that true?

**Comment by yairmov** [comment]
Thanks for raising this issue. I see how it can be confusing.
**tl;dr: For Resnet models use: GammaFlopsRegularizer**.

The rational:
Morphnet works by regularizing the "norm" of different operations in the graph. For a convolution we target the norm of each activation channel and try to reduce that. We do that by using the Group Lasso technique, which "groups" all the values in an activation channel, and tries to reduce all those values to 0. If all the values in a channel are 0, it means the network is not using it, and we mark it as "dead".

However, many modern CNNs (for example, Renset), use [Batch Norm](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm) (BN) after the convolution. For an activation channel X, the value after BN is:
```python
def batch_norm(x, gamma):
  return (x - mean(x)) / std(x) * gamma
```
This causes an issue if you are trying to regularize the activation:  The weight matrix is scale invariant!
```python
batch_norm(W*x) == batch_norm((W/2.0)*x)
```

There is, however an easy solution to this problem. Gamma! The gamma at the end of the BN operation becomes the scale of the OP. We can now target its absolute value as way to regularize the activation.

The "Gamma" in `GammaFlopsRegularizer` refers to this gamma variable in batch norm. 

So the easy rule of thumb is: If you have Batch Norm in your model, use `GammaFlopsRegularizer`, if your model uses convolutions without BN use `GroupLassoFlopsRegularizer`

**Comment by Aspirinkb** [comment]
So I should use `GammaFlopsRegularizer` if a Resnet model has one or more convolution layers followed by BN layers and some convolution layers not.

**Comment by yairmov** [comment]
If you use `GammaFlopsRegularizer` then only the BN layers will be regularized. If you use `GroupLassoFlopsRegularizer` only the non BN convs layers will be regularized. So I would say the decision depends on where most of the FLOPs in your model are.

In most use cases, the vast majority of conv layers have BN and only one or two do not (last layer), so I would use `GammaFlopsRegularizer`. If your use case is more customized and less standard, please describe it in more detail here and we can provide better advice.

Cheers,

**Comment by erichhhhho** [comment]
@yairmov In the paper:

> There are multiple ways to group
them, and in the results presented in this work we use the
L∞ norm - the maximum of the |γ|’s in the group.

So I was wondering where in the code reflect this point?
